### ISSUE TITLE
pkgmatch: Find R Packages Matching Either Descriptions or Other R Packages

### ISSUE THREAD

**Issue by mpadge** [ropensci/software-review#671]
**Date accepted:** 2026-06-23

Submitting Author Name: Mark Padgham
Submitting Author Github Handle: @mpadge
Repository:  https://github.com/ropensci-review-tools/pkgmatch
Version submitted: 0.4.2
Submission type: Standard
Editor: @MargaretSiple-NOAA
Reviewers: @agricolamz, @Selbosh

Archive: TBD
Version accepted: TBD
Language: en

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: pkgmatch
Title:  Find R Packages Matching Either Descriptions or Other R Packages
Version: 0.4.2
Authors@R: c(
    person("Mark", "Padgham", , "mark.padgham@email.com", role = c("aut", "cre"),
           comment = c(ORCID = "0000-0003-2172-5265")),
    person("Davis", "Vaughan", , "davis@posit.co", role = c("ctb"))
    )
Description: Find R packages matching either descriptions or other R packages.
License: MIT + file LICENSE
URL: https://docs.ropensci.org/pkgmatch/,
    https://github.com/ropensci-review-tools/pkgmatch
BugReports: https://github.com/ropensci-review-tools/pkgmatch/issues
Imports:
    brio,
    checkmate,
    cli,
    curl,
    dplyr,
    fs,
    httr2,
    memoise,
    pbapply,
    Rcpp,
    rvest,
    tibble,
    tidyr,
    tokenizers,
    treesitter,
    treesitter.r,
    vctrs
Suggests:
    gert,
    hms,
    httptest2,
    jsonlite,
    piggyback,
    pkgbuild,
    rappdirs,
    roxygen2,
    testthat (>= 3.0.0),
    withr,
    knitr,
    rmarkdown
LinkingTo:
    Rcpp
Depends: R (>= 3.5.0)
NeedsCompilation: yes
Encoding: UTF-8
Language: en-GB
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.3.2
Config/testthat/edition: 3
VignetteBuilder: knitr
```

## Scope

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [x] data retrieval
	- [ ] data extraction
	- [ ] data munging
	- [ ] data deposition
    - [ ] data validation and testing
	- [ ] workflow automation
	- [ ] version control
	- [ ] citation management and bibliometrics
	- [x] scientific software wrappers
	- [ ] field and lab reproducibility tools
	- [ ] database software bindings
	- [ ] geospatial data
	- [ ] text analysis
  - [x] rOpenSci tools

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):

Data retrieval, because the package includes code to generate language model (LM) embeddings from all R packages retrieved from both CRAN and rOpenSci package repositories. Wrapper because LM embeddings are generated by wrapping interface to [ollama](https://ollama.com) software. Plus I've inserted a new, one-off category of "rOpenSci tools" for internal, staff-curated packages.

-   Who is the target audience and what are scientific applications of this package?

Beyond internal rOpenSci use, target audiences are (1) entirely general audience of those interested in searching R packages using either text or code input, and (2) package developers, who can use this package to identify similar packages or functions to code they might be working on.

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

No, not at all. There are to my knowledge two other R packages for interfacing with LMs: [tidyllm](https://github.com/edubruell/tidyllm) and [elmer](https://github.com/tidyverse/elmer). Both of these are general interfaces to LM API endpoints, while this package specifically uses LM outputs to identify best-matching packages.

-   (If applicable) Does your package comply with our [guidance around _Ethics, Data Privacy and Human Subjects Research_](https://devguide.ropensci.org/policies.html#ethics-data-privacy-and-human-subjects-research)?

Not applicable.

-   If you made a pre-submission inquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

-   Explain reasons for any [`pkgcheck` items](https://docs.ropensci.org/pkgcheck/) which your package is unable to pass.

## Technical checks

Confirm each of the following by checking the box.

- [x] I have read the [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).
- [x] I have read the [author guide](https://devdevguide.netlify.app/authors-guide.html) and I expect to maintain this package for at least 2 years or to find a replacement.

This package:

- [x] does not violate the Terms of Service of any service it interacts with.
- [x] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme).
- [x] includes [documentation with examples for all functions, created with roxygen2](https://ropensci.github.io/dev_guide/building.html#documentation).
- [x] contains a vignette with examples of its essential functions and uses.
- [x] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [x] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage.

## Publication options

- [x] Do you intend for this package to go on CRAN?
- [ ] Do you intend for this package to go on Bioconductor?

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:

<details>
<summary>MEE Options</summary>

- [ ] The package is novel and will be of interest to the broad readership of the journal.
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.org/code-of-conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by emilyriederer** [comment]
@ropensci-review-bot assign @MargaretSiple-NOAA as editor

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot seeking reviewers

**Comment by MargaretSiple-NOAA** [comment]
(just a note: I put 'seeking reviewers' before putting in my editor checks but I have not forgotten them. I started them today and will continue tomorrow.)

**Comment by MargaretSiple-NOAA** [comment]
### Editor checks:

- [X] **Documentation**: The package has sufficient documentation available online (README, pkgdown docs) to allow for an assessment of functionality and scope without installing the package. In particular,
    - [X] Is the case for the package well made?
    - [X] Is the reference index page clear (grouped by topic if necessary)?
    - [X] Are vignettes readable, sufficiently detailed and not just perfunctory?
- [X] **Fit**: The package meets criteria for [fit](https://devguide.ropensci.org/policies.html#package-categories) and [overlap](https://devguide.ropensci.org/policies.html#overlap).
- [ ] **Installation instructions:** Are installation instructions clear enough for human users? - *Yes but see notes below.*
- [ ] **Tests**: If the package has some interactivity / HTTP / plot production etc. are the tests using [state-of-the-art tooling](https://devguide.ropensci.org/building.html#testing)? *Yes but see notes below. I think I just can't fully evaluate this until I can install docker and ollama on my personal computer and test the functions*
- [X] **Contributing information**: Is the documentation for contribution clear enough e.g. tokens for tests, playgrounds?
- [X] **License:** The package has a CRAN or OSI accepted license.
- [X] **Project management**: Are the issue and PR trackers in a good shape, e.g. are there outstanding bugs, is it clear when feature requests are meant to be tackled?
---

#### Editor comments
Nice package, @mpadge ! This will be useful for anyone working on package development or review. I wrote a few notes on what I saw during the editor check process-- they're longer than I usually write, but I figure these will probably make the reviewers' lives easier so it doesn't hurt to mention them early.

A few notes of mine from the editor check process:
- The pkgdown page seems to indicate that the user can set the corpus to rOpenSci or to CRAN, but some functions only provide search results for rOpenSci, e.g., `pkgmatch_bm25()` requires the corpus to be rOpenSci-related. Some clarity about which functions apply to which package corpi (?) might be helpful.
- I got 1 note from running `devtools::check()`: "Package suggested but not available for checking: 'piggyback'" -- I don't worry about these too much just wanted to flag it just in case.

- The ollama connection gave me some trouble at the start, so I imagine future users might struggle on this. I recommend revising the documentation a little bit to make it more obvious what needs to happen with ollama before users can get started: a) Consider adding a sentence to the top of the "Get started" article that indicates that docker and ollama installs need to happen first and foremost. Here, I would link to the "ollama" article. b) Change the title of the "ollama" article to say something like "Before you begin: ollama" or "Before you begin: ollama installation", so that when people are browsing, it's very clear to them that they'll need to read that article first. 
- The beginning of the [Docker section](https://docs.ropensci.org/pkgmatch/articles/ollama.html#docker) in the ollama article should indicate that users need to have Docker installed as well. I learned from this process that my institution actually doesn't have a Docker license! This may impede editors/reviewers who work at NOAA. It sounds like they're working on getting a license so that may change in the future, so that will probably change in the future.

- I had some trouble getting test coverage my usual way (`covr::package_coverage()` fails with an error about a temp file)... but `devtools::test()` works great). Some tests are failing but I'm pretty sure that's because I don't have ollama properly installed on my machine. It might be nice to have a one-line check in one of the early vignettes that can show people whether they are missing any of the components need to run the fns.


---

**Comment by mpadge** [comment]
Thanks @MargaretSiple-NOAA for really useful feedback! The issue linked above has details of changes made in response. I'd say the most important of those for reviewers of this package is that I've added [`tests/README`](https://github.com/ropensci-review-tools/pkgmatch/tree/main/tests) explaining that tests can be run without having `ollama` installed or running anywhere.

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot Add @agricolamz as reviewer

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot set due date for @agricolamz to 2025-01-31

**Comment by MargaretSiple-NOAA** [comment]
Hi @mpadge -- I was just revising my editor checks above after your revisions, and everything is looking good except some issues I had with `devtools::test()` - these are documented in the issue I logged on the pkgmatch github page (above). I don't know what they mean but I think we should have a resolution before reviewers are running similar checks.

**Comment by mpadge** [comment]
Thanks @MargaretSiple-NOAA, I've fixed the issue linked above, so `devtools::test()` should now work for everybody.

**Comment by MargaretSiple-NOAA** [comment]
Thank for for addressing that , @mpadge ! The tests now run without issues. 

Two more quick things:
- `pkgmatch_similar_pkgs(..., corpus = "cran")` always gives a warning `Error in matrix(emb[[what]], nrow = nrow, ncol = npkgs) : 
  non-numeric matrix extent`. This renders fine on the pkgdown page [here](https://docs.ropensci.org/pkgmatch/articles/pkgmatch.html) so I suspect it is a computer-specific issue. Everything works as expected when `corpus = "ropensci"`. 
- Several of the tests fail and trace back to `parseNamespaceFile(pkg_name, package.lib = lp)`. I think it's fine to fix this when reviews are in but wanted to give you a heads-up while I'm still searching for a Reviewer 2.

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot add @Selbosh as reviewer

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot Set due date for @Selbosh to 2025-02-21

**Comment by maurolepore** [comment]
Dear @mpadge this is to mark the start of my EiC rotation. I'm reviewing all open issues and noting what I see.

I see one review is due about now and the second one on Feb 21st. It all looks good to me. I'll step back. Thanks!

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot Set due date for @agricolamz to 2025-02-21

**Comment by Selbosh** [comment]
@mpadge Having some issues getting `ollama` to cooperate on Ubuntu; unfortunately this also causes the package to give a cryptic error message.

```r
pkgmatch_similar_pkgs(input)
```
```
Error in names(out) <- nms : 
  'names' attribute [1] must be the same length as the vector [0]
```
Found some documentation [here](https://pypi.org/project/ollamac/) so will try to figure it out and get the review done soon.

**Comment by Selbosh** [comment]
I think I found one half of the the problem: `set_ollama_url` does not modify the system environment variables, so the system call to `ollama list` within `ollama_check()` will fail depending on how you have set up your shell environment variables (e.g. if you `export OLLAMA_HOST=localhost:8888` but then open R in another session, then the exported variable is lost).

In addition, `ollama_check()` fails and gives the error above if `ollama` is running, but without any models downloaded (it's trying to iterate over a list of length zero, I guess?)

So the current sequence, after setting `ollama serve` running on `localhost:8888` in another shell and manually pulling the two embeddings in the vignette, is:

```r
Sys.getenv('OLLAMA_URL')
## [1] ""
library(pkgmatch)
ollama_check()
## Error: could not connect to ollama app, is it running?
## Error in out[[1]] : subscript out of bounds
## In addition: Warning message:
## In system("ollama list", intern = TRUE) :
##  running command 'ollama list' had status 1
set_ollama_url('127.0.0.1:8888')
ollama_check()
## Error: could not connect to ollama app, is it running?
## Error in out[[1]] : subscript out of bounds
## In addition: Warning message:
## In system("ollama list", intern = TRUE) :
##  running command 'ollama list' had status 1
Sys.setenv(OLLAMA_HOST = '127.0.0.1:8888')
ollama_check()
## [1] TRUE
```

But then I still do not get much farther, because
```r
pkgmatch_similar_pkgs(input)
## Error in `httr2::req_perform()`:
## ! Failed to perform HTTP request.
## Caused by error in `curl::curl_fetch_memory()`:
## ! Couldn't connect to server [localhost]: Failed to connect to localhost port 11434 after 1 ms: Connection refused
## Run `rlang::last_trace()` to see where the error occurred.
```
which implies that there is least one other place that `11434` is hard-coded. A quick search found

https://github.com/ropensci-review-tools/pkgmatch/blob/21c1e138535f86de6501ec8549f7c8a6f6ac4bf9/R/embeddings.R#L240

which unfortunately means I can't proceed on Ubuntu without fixing this bug first.

**Comment by Selbosh** [comment]
## Package Review

Hello @mpadge and @MargaretSiple-NOAA! Thank you giving me the opportunity to review the package. It is my first package review, and I am still learning the process, so I hope that my feedback is useful. As mentioned above, I had a couple of issues getting started, but have found some fixes that should with any luck make it easier for others who encounter similar problems.

Please let me know if I can clarify anything. Here are my comments:

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [ ] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions
- [x] **Examples:** (that run successfully locally) for all exported functions
- [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software have been confirmed.
- [x] **Performance:** Any performance claims of the software have been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing: 5

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

#### Documentation

In the `README`, the statement of need should be more prominent; it's not obvious who the target audience of the package is or why the package is needed. It would be great if you could explain early on, clearly, what problem `pkgmatch` solves, maybe giving examples of use cases from the start.

I think that you could improve the documentation by making it more self-contained and making it easier for the user to find information quickly. Some information that should be in the `README` is only in a vignette, and some information that should be at the start of one vignette is instead left as a link to another vignette. The dependence on LLMs is also somewhat buried: this could be a make-or-break feature for many users, but it's only mentioned at the bottom of the Installation section. I can imagine it could be a selling point that `pkgmatch` exploits LLMs, but it could also be a deal-breaker for some users either due to technical limitations or ethical concerns.

The package relies on `ollama` but it's not explained in the README what this is or why the package should depend on it, or whether you can use some functions without it. There is a function called `ollama_check()` that ostensibly checks for and installs `ollama`: add this to the `README`, as it seems like the first thing you would want to run after loading the package! "Procedures for setting that up are described in a separate vignette" is not super helpful without a direct link to or giving the name of said vignette! Ideally I'd put much of this information in the README, because it could be a decider on whether somebody can/will install the package in the first place. Given we are talking about local LLMs here, not everybody will know what their computational demands are, so it would be helpful to give some general indication of system requirements in the 'Before you begin' vignette and maybe in the README.

Generally speaking, I think the documentation assumes a bit too much knowledge on the part of the user, by not motivating the task clearly enough, not explaining what some functions do or what some terms mean. There are one or two functions documented in terms of what they are _not_ rather than what they _are_, which can be a bit confusing (e.g. `pkgmatch_bm25_fn_calls()`). Some function descriptions (e.g. `pkgmatch_bm25()`) simply comprise an external Web link, which is not ideal for offline use, nor for users who want to know what a function does without having to click around. For the same functions, the current Titles in the documentation are too long or specific and would be better off as Descriptions, and the current Descriptions as Details or Notes.

The purpose of some of the vignettes is unclear. For example, "Good-enough practices for language model packages": who is this for, exactly? Why does it need to be a vignette? Why do "Data caching and updating" or "Why local LMs?" need their own vignettes, instead of being included as sections in a larger general guide? This information is surely useful, but it could be merged into a single vignette and/or the `README`, maybe with a dedicated vignette for developers, to better signpost the user to the information they need (or don't need).

#### Functionality

I was able to install the package (v0.4.3.018) first try on WSL2 (Ubuntu on Windows 11) and didn't face any issues with building vignettes or `R CMD CHECK`. All tests passed successfully.

Not really sure where to start, I decided to run the example in the `README`:
```r
pkgmatch_similar_pkgs(input)
```

This threw an error:
```
Error in names(out) <- nms : 
  'names' attribute [1] must be the same length as the vector [0]
```

I then remembered that I hadn't started the `ollama` server, so I go to the command line and run `ollama serve`, to get
```
Error: listen tcp 127.0.0.1:11434: bind: address already in use
```
This might be quite esoteric but could be worth mentioning in the vignette. A bit of detective work suggests that the error message above is in fact coming from `ollama_check()`, which when run throws exactly the same error. This left me a bit stuck, due to a couple of issues in the error handling, and due to [a bug in the `embeddings.R` file](https://github.com/ropensci-review-tools/pkgmatch/issues/106). I made a local patch to fix it on my end and was able to proceed from here.

Once I got the package working, the benefits of it seem clear! I can see this being immensely useful to package developers and advanced users who are looking for existing functionality, either for specialist tasks or trying not to reinvent the wheel while developing their own packages. However, as described above, the documentation comes across as being aimed at people who are already familiar with the package, rather than to new users who are trying to get it to work for the first time.

Another interface issue I note is that with most function calls, there is a slight delay and we see that data is being downloaded in the background:
```r
pkgmatch_similar_pkgs (".", corpus = "cran")
##  [100%] Downloaded 231213455 bytes...
##  [100%] Downloaded 36834164 bytes...
##  [100%] Downloaded 3948804 bytes...
##  [100%] Downloaded 13016426 bytes...
```
but what data is being downloaded, why and where to? Is it the corpus? Is it the embeddings? Maybe we should ask the user if they'd like to use up their bandwidth and disk space on this, or reassure them that it's going to a temporary directory and/or can easily be cleaned out at the end of the session? Which functionality will work offline, if any? This could be made a bit clearer on the first use.

The `browse = TRUE` function threw a further error for me, possibly related to using VS Code.
```r
pkgmatch_similar_fns (input, browse = TRUE)
# VSCode WebView only supports showing local http content.
# Opening in external browser...
# Browsing https://docs.ropensci.org/charlatan/reference/SequenceProvider
# Error in vapply(urls, utils::browseURL, integer(1L)) : 
#   values must be length 1,
#  but FUN(X[[1]]) result is length 0
```
However, the help page successfully opened in the browser, so I'm not sure what the issue is here. I was able to run other examples successfully locally, though nearly all of them are 'Don't run' because of the external dependencies; I wonder if there is the possibility to embed some toy embeddings and `system.file`s in the package?

#### Other comments

In the examples and `README`, you have spaces between function names and the brackets, e.g. `fun ()` instead of `fun()`, which is not recommended in [most style guides](https://devguide.ropensci.org/pkg_building.html#code-style).

Overall I think this is a potentially great package with real potential, but the main thing holding it back is selling to the user what it's for and giving an easy way to get started. There are also one or two bugs or edge cases that thwarted me personally, but might not come up for other users, and so this might have coloured my impressions of the package. To do this, I'd suggest filling in the gaps and maybe focus on having a vignette with an extended use case that shows off the package's capabilities.

In the long run, I'd suggest clearly marking out which functionality is dependent on which external tools, as then a new user might be able to explore some features even if they have issues with others or do not want to set them up yet.

**Comment by mpadge** [comment]
Thank you very much @Selbosh for your review. You've really immediately picked up the big weakness of package design thus far: It was pulled together for a very specific use case, with documentation mostly hobbled together around that narrow vision, rather than me having thought nearly as much as I should have about general usage. In short: I agree with all of your concerns, and will address them hopefully within the coming fortnight. You should see several new issues in the repo linked immediately below here.

**Comment by Selbosh** [comment]
Thank you @mpadge, I am glad if my feedback can be helpful and hope it didn't come across as sounding too harsh.

**Comment by MargaretSiple-NOAA** [comment]
Thank you @Selbosh for your thorough review and @mpadge for the quick response!

**Comment by mpadge** [comment]
Thanks @MargaretSiple-NOAA. Just to let you know, I plan on waiting for second review before tackling the issues from @Selbosh's review, to ensure both reviews pertain to the package in a similar state.

**Comment by agricolamz** [comment]
Hi, all, sorry but my review will be late for more two days, but I'm planning to finish it on Sunday.

**Comment by MargaretSiple-NOAA** [comment]
No problem George-- I'll adjust the due date just for our records but Sunday is fine.

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot set due date for @agricolamz to 2025-02-23

**Comment by agricolamz** [comment]
## Package Review

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions
- [x] **Examples:** (that run successfully locally) for all exported functions
- [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software have been confirmed.
- [x] **Performance:** Any performance claims of the software have been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing: 9

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

I was avoiding reading the review by @Selbosh, and now I see that some of my comments are the same.

## Major comments

- I am not happy with the side effects of the package. Here is my first run of the `pkgmatch_similar_pkgs()` function:

```
> library(pkgmatch)
> input <- "
+ Packages for analysing evolutionary trees, with a particular focus
+ on visualising inter-relationships among distinct trees.
+ "
> pkgmatch_similar_pkgs(input)
 [100%] Downloaded 3497647 bytes...
 [100%] Downloaded 1076837 bytes...
Would you like to download it now (y/n) ? n
Would you like to download it now (y/n) ? n
Error in `httr2::req_perform()`:
! HTTP 404 Not Found.
Run `rlang::last_trace()` to see where the error occurred.
Warning messages:
1: ollama model [jina/jina-embeddings-v2-base-en] is not installed. 
2: ollama model [ordis/jina-embeddings-v2-base-code] is not installed. 
```

First, it seems that the package quietly downloads something, and I don’t know what it is or where it is stored (now I know that it is in `/home/agricolamz/.cache/R`). Next, the function asked me whether “I would like to download it now?”. But I see that the function already downloaded something... Why did the package suddenly become so kind on the third and the fourth download?

After I installed the `ollama` models `jina/jina-embeddings-v2-base-en` (274 MB) and `ordis/jina-embeddings-v2-base-code` (323 MB) and played around, I decided to run the function using the `corpus = 'cran'` argument and got even bigger downloads:

```
[100%] Downloaded 232067039 bytes...
[100%] Downloaded 37056488 bytes...
```

I don't think that 273 Mb will cause any damage, but I expect that most users will disagree that silence implies consent.

- Another problem that I have a feeling that package does not work (the way I expect).

First I decided to check some popular packages

```{r}
$ pkgmatch_similar_pkgs(corpus = "cran",
                      "Package that works with dates and times in the tidy format")

[1] "iso8601"         "rebus.datetimes" "datetime"        "tibbletime"     
[5] "datefixR" 
```

`lubridate` is number 15. I decided to check, whether the result is LLM motivated, so I ran 6 possible proportions of `lm_proportion`:

![Image](https://github.com/user-attachments/assets/742f392b-c5f7-41e3-b529-bebd0fb32e8c)

![Image](https://github.com/user-attachments/assets/a2f22dbe-f939-470b-bd67-4417c1e5666d)

![Image](https://github.com/user-attachments/assets/d2b80334-edfa-410e-b9ab-fa387d29385b)

![Image](https://github.com/user-attachments/assets/ef605001-d65a-4b95-9ba4-16a3792db4a1)

![Image](https://github.com/user-attachments/assets/3f72fc76-0cdc-4d1d-ac96-5cfb32b4443c)

![Image](https://github.com/user-attachments/assets/e86f14d3-56a7-437f-9100-5968280c5af3)

As you can see from the plots, my expectations were not met. That was striking for me, since I based my prompts on the documentation phrasing. It is possible that my function call meets this 8,096 token limit, so the model took the latest part of the package, omitting the very first entries. But this is just a hypothesis. I must admit: the suggested packages sometimes are really close to the prompt query. If you want to fix this, I would suggest merging information from LLM and BM25 with information about the number of downloads or any other measure of the package's popularity.

I also test the `pkgmatch_similar_pkgs()` function with the lingtypology tar.gz and the result was suprising for me:

![Image](https://github.com/user-attachments/assets/31e011b2-6833-450d-9571-f9b19f972296)

As you see, the package correctly identifies the code, but for some reason, it struggles with the text.

- I disagree with your strategy of update of the pre-generated embeddings (see "Data caching and updating" vignette). I think that all updates should be done within the package itself, otherwise it is a high risk to get non-reproducible (even non-repeatable) result.

## Smaller comments/suggestions

- As you see, I have got an error on the first run. So I would propose to add a possibility to kindly refuse downloading something (without throwing an error).
- It looks like you are using two specific ollama models in the package; however, if I have an ollama on my computer, it is highly probable that I already downloaded some models (e.g. llama), and the embeddings from them could be useful. So I’d propose to make it possible to use `ollamar::embeddings()` and call some models that already present on the computer.
- With all due respect to the ROpenSci community, I think that corpus = "cran" should be the default option.
- The `browse = TRUE` argument seems to be not working in RStudio. It opens the link for the first package and throws an error:

```
Error in vapply(urls, utils::browseURL, integer(1L)) : 
  values must be length 1,
 but FUN(X[[1]]) result is length 0
```

- It is not a problem at all, rather a question. I see through all documentation a space between the function name and the bracket. Is it a new code style that I am not aware of?

- Another side effect that I found is that the output of the `pkgmatch_similar_pkgs()` function is a dataframe and not a vector as I expected from the function's behavior. As I realized, there is a special print method (it is documented in the documentation). I would propose to have an additional argument that makes the function return a vector or the whole sorted dataframe. I am suggesting this because even the datafarame output is of a class pkgmatch that for some reason makes it impossible to merge multiple instances of the function output (e.g. gathered with the different lm_proportion value).

- I find the "Good-enough practices for language model packages" article from the documentation unnecessary. The contents of the "Why local language models (LMs)?" may can be devided between "README.md" and "Get started" page.

- Why is there no `lm_proportion` argument in the `pkgmatch_similar_fns()` function?

- You forgot the ending bracket in `fs::path(rappdirs::user_cache_dir(), "R", "pkgmatch"` in the "Data caching and updating" vignette.

**Comment by agricolamz** [comment]
Dear @mpadge, it is my pleasure! I hope my comments will help to improve the package!

**Comment by MargaretSiple-NOAA** [comment]
Thank you @agricolamz and @Selbosh for two great and thorough reviews.  @mpadge  I'll check back in after the 10th.

**Comment by mpadge** [comment]
@MargaretSiple-NOAA @Selbosh @agricolamz I've started updating the package in response to the really helpful reviews. You can see several issues listed above. I hope to finish them within the next couple of weeks, and will then write a full response here. Thank you all 👍 😄

**Comment by maurolepore** [comment]
Deal all I'm wrapping my EiC rotation and before I "go" I'd like to thank you very much for your work!

**Comment by MargaretSiple-NOAA** [comment]
Thank YOU @maurolepore for keeping us all on track! 

@mpadge , FYI, I am going into be at sea for field work May 23-July 11 this year so if you are still working on it for a while longer, we may have a period of inactivity during which we could put it on hold or [some other option that I don't know yet].

**Comment by agricolamz** [comment]
As a field linguist myself, I wish you good luck!

**Comment by mpadge** [comment]
Thanks both. In that case I'll endeavour to get my responses back before you head off. I'll aim to do that quickly enough to finish this before May 23rd. If not, we'll maybe consider [some other option that I don't know yet] 😄 ... but hopefully not 🚀

**Comment by mpadge** [comment]
@MargaretSiple-NOAA Update: I'm not quite going to get my responses done by May 23rd. If I may, please let me know what you think of this as a plan:

- I get my responses done and posted here by end of this month of May
- I'll ping both @Selbosh and @agricolamz for their thoughts, feedback, and judgement
- We back-and-forth during month of June while you're away
- Hopefully by mid-July when you return, everything will be ready to approve.

Let me know whether that sounds good, and David and George, hopefully you'll both have a small amount of time in June to have a look over my changes. Thanks all!

**Comment by agricolamz** [comment]
Hi all, I don't think that I will be available through the June, but I will try to react in my spare time. Good luck in a field @MargaretSiple-NOAA!

**Comment by MargaretSiple-NOAA** [comment]
Hi @mpadge -- I am at sea currently and I realize I haven't kept up on my check-ins. Shall we put this on hold until July 11, when I return from this survey? That will give you some additional time to get through these very thorough and helpful comments from both reviewers.

**Comment by mpadge** [comment]
Thanks @MargaretSiple-NOAA, yes, I think that's a good idea. Updates in the meantime:

- I failed to keep my [previous promise](https://github.com/ropensci/software-review/issues/671#issuecomment-2894480752) to get responses back in time, for which I can only apologise
- i have nevertheless been implementing many improvements, most notably through addressing the concerns of @agricolamz about results not matching expectations.
- Documentation and vignettes have been improved and greatly extended.

I'm confident I'll be able to post something here before July 11. Thank you all for your patience.

**Comment by mpadge** [comment]
I am deeply indebted to both @Selbosh and @agricolamz for your extremely helpful reviews. With due apology for my response having taken so long, please note that this was largely because the package has been very extensively rewritten, restructured, and improved in response to your feedback.

Please note also that the following responses were largely written as "mere" technical responses, and so are largely devoid of expressions of gratitude or appreciation. Rest assured: all comments were enormously appreciated, and changes made in response have really helped to improve the package enormously, for which I am very grateful. Thank you both! I'll post separate responses here, and for ease of subsequent discussion prefix each one with unique identifiers.

# Responses to @Selbosh

## Documentation

#### DS01

> In the README, the statement of need should be more prominent; it's not obvious who the target audience of the package is or why the package is needed. It would be great if you could explain early on, clearly, what problem pkgmatch solves, maybe giving examples of use cases from the start.

Addressed via https://github.com/ropensci-review-tools/pkgmatch/issues/109 and https://github.com/ropensci-review-tools/pkgmatch/issues/117

README now starts with explicit description of LLM usage, and ["Get started" vignette](https://docs.ropensci.org/pkgmatch/articles/pkgmatch.html) starts with a longer description of what the package does and why. That vignette also proceeds straight to a [_What does the package do?_ section](https://docs.ropensci.org/pkgmatch/articles/pkgmatch.html#what-does-the-package-do).

---

#### DS02

> I think that you could improve the documentation by making it more self-contained and making it easier for the user to find information quickly. Some information that should be in the README is only in a vignette, and some information that should be at the start of one vignette is instead left as a link to another vignette.

Addressed mainly via https://github.com/ropensci-review-tools/pkgmatch/issues/111, which extensively restructured the vignettes. The most important improvement is the list at the [end of the REAMDE](https://docs.ropensci.org/pkgmatch/index.html#package-vignettes) with brief descriptions of the contents and purposes of all vignettes. As explained [in that issue](https://github.com/ropensci-review-tools/pkgmatch/issues/111#issuecomment-3051817461),

---

#### DS03

> All individual vignettes have been retained (and an extra one added via #118), because I hope this should enable people to more readily find what they need, rather than having to parse sections of longer vignettes

> The dependence on LLMs is also somewhat buried: this could be a make-or-break feature for many users, but it's only mentioned at the bottom of the Installation section. I can imagine it could be a selling point that pkgmatch exploits LLMs, but it could also be a deal-breaker for some users either due to technical limitations or ethical concerns.

That was a really important point, and has now been addressed by revising initial text of [main README](https://docs.ropensci.org/pkgmatch/index.html) to state that `pkgmatch` is,

> A tool that uses languages models to help find R packages.

All text changes are linked in [issue#117](https://github.com/ropensci-review-tools/pkgmatch/issues/117#issuecomment-3048200447). But most importantly, [issue#116](https://github.com/ropensci-review-tools/pkgmatch/issues/116) enables almost all functions to work without ollama, and issue warnings that results may be less accurate that those with ollama. The previous absolute need for ollama is thus effectively circumvented, which hopefully avoids the "deal-breaker" status.

---

#### DS04

> The package relies on ollama but it's not explained in the README what this is or why the package should depend on it, or whether you can use some functions without it. There is a function called ollama_check() that ostensibly checks for and installs ollama: add this to the README, as it seems like the first thing you would want to run after loading the package! "Procedures for setting that up are described in a separate vignette" is not super helpful without a direct link to or giving the name of said vignette! Ideally I'd put much of this information in the README, because it could be a decider on whether somebody can/will install the package in the first place. Given we are talking about local LLMs here, not everybody will know what their computational demands are, so it would be helpful to give some general indication of system requirements in the 'Before you begin' vignette and maybe in the README.

All done via above-linked https://github.com/ropensci-review-tools/pkgmatch/issues/116 and https://github.com/ropensci-review-tools/pkgmatch/issues/117. Initial bit of README now starts,

> This package relies on a locally-running instance of [ollama](https://ollama.com). Procedures for setting that up are described in a [separate vignette](https://docs.ropensci.org/pkgmatch/articles/B_ollama.html) (`vignette("ollama", package = "pkgmatch")`). Although some functionality of this package may be used without ollama, the main functions require ollama to be installed.

That vignette also now includes [this statement](https://docs.ropensci.org/pkgmatch/articles/B_ollama.html):

> Note that the [ollama](https://ollama.com) software is around 1GB, and its usage in `pkgmatch` requires downloading two additional language models of around 300MB each. Installation will thus use around 2GB of disk space (and will generally take quite some time). [ollama](https://ollama.com) will automatically run on GPUs if detected, but can also run (in somewhat slower mode) on CPUs.

---

#### DS05

> Generally speaking, I think the documentation assumes a bit too much knowledge on the part of the user, by not motivating the task clearly enough, not explaining what some functions do or what some terms mean. There are one or two functions documented in terms of what they are not rather than what they are, which can be a bit confusing (e.g. pkgmatch_bm25_fn_calls()). Some function descriptions (e.g. pkgmatch_bm25()) simply comprise an external Web link, which is not ideal for offline use, nor for users who want to know what a function does without having to click around. For the same functions, the current Titles in the documentation are too long or specific and would be better off as Descriptions, and the current Descriptions as Details or Notes.

All function documentation extensively revised and improved in response to this very valid criticism, mostly in https://github.com/ropensci-review-tools/pkgmatch/issues/110. See [example for `pkgmatch_bm25()`](https://docs.ropensci.org/pkgmatch/reference/pkgmatch_bm25.html), which now starts with an extended description of how this function actually works.

---

#### DS06

> The purpose of some of the vignettes is unclear. For example, "Good-enough practices for language model packages": who is this for, exactly? Why does it need to be a vignette? Why do "Data caching and updating" or "Why local LMs?" need their own vignettes, instead of being included as sections in a larger general guide? This information is surely useful, but it could be merged into a single vignette and/or the README, maybe with a dedicated vignette for developers, to better signpost the user to the information they need (or don't need). 

That previous "Good-enough practices" vignette has been completely removed. While I agree with the general concern that multiple vignettes may be initially more confusing that fewer, longer vignettes, I have nevertheless retained all individual ones, but added much clearer signposting via https://github.com/ropensci-review-tools/pkgmatch/issues/111. As stated above, I think that retaining multiple vignettes will help people more directly find relevant information, rather than having to wade through more extended ones. That then relies on them being sufficiently signposted, for which I hope that [the explanations in the README](https://docs.ropensci.org/pkgmatch/index.html#package-vignettes) do a good job.

----

## Functionality

#### DS07

> I was able to install the package (v0.4.3.018) first try on WSL2 (Ubuntu on Windows 11) and didn't face any issues with building vignettes or R CMD CHECK. All tests passed successfully.

> Not really sure where to start, I decided to run the example in the README:

> pkgmatch_similar_pkgs(input)

> This threw an error:

> Error in names(out) <- nms : 
>   'names' attribute [1] must be the same length as the vector [0]

> I then remembered that I hadn't started the ollama server, so I go to the command line and run ollama serve, to get

> Error: listen tcp 127.0.0.1:11434: bind: address already in use

> This might be quite esoteric but could be worth mentioning in the vignette. A bit of detective work suggests that the error message above is in fact coming from ollama_check(), which when run throws exactly the same error. This left me a bit stuck, due to a couple of issues in the error handling, and due to a bug in the embeddings.R file. I made a local patch to fix it on my end and was able to proceed from here.

I think all of those concerns have been addressed via https://github.com/ropensci-review-tools/pkgmatch/issues/112, https://github.com/ropensci-review-tools/pkgmatch/issues/113, and other issues cross-linked within those. There should generally be far fewer errors, especially as https://github.com/ropensci-review-tools/pkgmatch/issues/116 allows the package to work without ollama running at all.

---

#### DS08

> Once I got the package working, the benefits of it seem clear! I can see this being immensely useful to package developers and advanced users who are looking for existing functionality, either for specialist tasks or trying not to reinvent the wheel while developing their own packages. However, as described above, the documentation comes across as being aimed at people who are already familiar with the package, rather than to new users who are trying to get it to work for the first time.

Thank you for saying that! There's a whole new vignette with [_Example applications_](https://docs.ropensci.org/pkgmatch/articles/A_extended-use-case.html) which hopefully makes the general case even clearer and more convincing. (But see *DS13* below for remaining concerns I have with that vignette.)

---

#### DS09

> Another interface issue I note is that with most function calls, there is a slight delay and we see that data is being downloaded in the background:

> pkgmatch_similar_pkgs (".", corpus = "cran")
> [100%] Downloaded 231213455 bytes...
> [100%] Downloaded 36834164 bytes...
> [100%] Downloaded 3948804 bytes...
> [100%] Downloaded 13016426 bytes...

> but what data is being downloaded, why and where to? Is it the corpus? Is it the embeddings? Maybe we should ask the user if they'd like to use up their bandwidth and disk space on this, or reassure them that it's going to a temporary directory and/or can easily be cleaned out at the end of the session? Which functionality will work offline, if any? This could be made a bit clearer on the first use.

That has all been addressed via https://github.com/ropensci-review-tools/pkgmatch/issues/114, which means:

- An initial message is given saying how many datasets need to be downloaded
- Downloads then give name of the download, as well as count (`1 / 10`, for example)
- Information is given about _where_ these are cached, and also that this cache directory may be safely removed at any time.

---

#### DS10

> The browse = TRUE function threw a further error for me, possibly related to using VS Code.

> pkgmatch_similar_fns (input, browse = TRUE)
> VSCode WebView only supports showing local http content.
> Opening in external browser...
> Browsing https://docs.ropensci.org/charlatan/reference/SequenceProvider
> Error in vapply(urls, utils::browseURL, integer(1L)) : 
>   values must be length 1,
>  but FUN(X[[1]]) result is length 0

I think that should have all been fixed. Linking functions to package URLs is a bit of a dark art, because `.Rd` files actually specify a "topic" which may or may not match on to actual function names. I expect issues may still arise when trying to open pages from `pkgmatch_similar_fns()`, but am confident it's at a generally useable state from which issues can be invidiually resolved if an when they arise.

---

#### DS11

> However, the help page successfully opened in the browser, so I'm not sure what the issue is here. I was able to run other examples successfully locally, though nearly all of them are 'Don't run' because of the external dependencies; I wonder if there is the possibility to embed some toy embeddings and system.files in the package?

That was a great idea, and has been done exactly as suggested with the help of [a `generate_pkgmatch_example_data()` function](https://docs.ropensci.org/pkgmatch/reference/generate_pkgmatch_example_data.html). See active example in use at https://docs.ropensci.org/pkgmatch/reference/pkgmatch_similar_pkgs.html#ref-examples.

---

#### DS12

> In the examples and README, you have spaces between function names and the brackets, e.g. fun () instead of fun(), which is not recommended in most style guides.

I know, and apologise for any cognitive challenge in visually parsing my code style. It's called [the `spaceout` style](https://github.com/ropensci-review-tools/spaceout), and explained [in 'CONTRIBUTING.md'](https://github.com/ropensci-review-tools/pkgmatch/blob/main/CONTRIBUTING.md#code-style).

---

#### DS13

> Overall I think this is a good idea for a package with real potential, but the main thing holding it back is selling to the user what it's for and giving an easy way to get started. To do this, I'd suggest filling in the gaps and maybe focus on having a vignette with an extended use case that shows off the package's capabilities.

That's really useful feedback, thank you. The ["extended use case" vignette](https://docs.ropensci.org/pkgmatch/articles/A_extended-use-case.html) has been written. I still worry that I'm a bit too technically close to everything here to clearly see the proverbial forest, so would appreciate any specific further feedback you might be able to provide in order to better address your concerns here. I belive reasonably firmly that I have satisfactorily addressed all review concerns with the singular exception of this one, which still worries me a bit. Any and all further help will be greatly appreciated!

---

#### DS14

> In the long run, I'd suggest clearly marking out which functionality is dependent on which external tools, as then a new user might be able to explore some features even if they have issues with others or do not want to set them up yet.

I trust this has been addressed via https://github.com/ropensci-review-tools/pkgmatch/issues/116, which allows (almost) everything to work without external tools (ollama).

**Comment by agricolamz** [comment]
Hi, I didn't have much time to react yet, but I wanted to note that it make sense to add some motivation for using your package comparing it to the existing tools:

1) https://r-packages.io/
2) LLMs themselfs

**Comment by MargaretSiple-NOAA** [comment]
Sorry for the delay on this package all-- I am about to do some bookkeeping so my apologies for all the notifications you're about to get!

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/671#issuecomment-2654565738 time 5

**Comment by MargaretSiple-NOAA** [comment]
Hi @mpadge, thank you for the revisions based on these great comments. I think the package has benefitted greatly from both reviews. 

I have a couple of editing questions/comments while we await approvals:
- Can you use that 'submit response' command to the ropensci-review-bot so it officially logs both your responses to reviewers? I've seen folks skip this part but I want to be thorough.
- I really like your vignette **Example applications**. This seems like a good place for @agricolamz 's suggested packages/tools that might do comparable searching tasks (r-packages dot io, searches with broader multipurpose LLMs).
- I have found some clarity/usability edits that I think I would recommend to the vignettes (some are copy edit-adjacent but still impact the interpretation). I'm happy to handle those (or not) however you'd like.

@Selbosh and @agricolamz -- please look over @mpadge's changes and if you are satisfied with the revisions, post your approvals (or further remarks).

**Comment by mpadge** [comment]
@Selbosh Thanks again for your helpful comments.

- The difficulties with initial downloading of ollama models were something I hadn't really tested or debugged at all, so your insights there helped a lot. All fixed in https://github.com/ropensci-review-tools/pkgmatch/issues/203.
-  The `piggyback` dependency has been moved from _Suggests_ to _Imports_, as you suggested
- Finally, the information when downloading pre-computed embedding data has been updated and made more informative and verbose. It now looks like this:

```
This function requires data to be downloaded.
Data will be downloaded to `~/.cache/R/pkgmatch`.
This directory may be safely deleted at any time.
See the pkgmatch 'Data caching and updating' vignette for details.
Data for the ropensci corpus comprised 2 files totalling around 6.4MB.
Do you want to proceed (y/n)?
```

**Comment by mcsiple** [comment]
Hi all -- I am writing from my academic GitHub here so I can follow this thread. FYI I am still keeping tabs on this review despite the shutdown.

**Comment by MargaretSiple-NOAA** [comment]
Hi @mpadge , this looks awesome! @agricolamz and @Selbosh I think these updates are good -- if you approve them can you approve with the [template](https://github.com/ropensci/software-review/issues/671#issuecomment-3697233936)? If not, we can continue to hash it out here. If possible, I would like to move this along before I leave for summer survey season at the end of May. Thank you all for your hard work and patience.

**Comment by agricolamz** [comment]
Hi all, I'll try to review the result at the end of April, when I'll be back from the field trip.

**Comment by maurolepore** [comment]
Hi @mpadge, this is to mark the start of my new EiC rotation.

Thanks @Selbosh and @agricolamz for the reviews, and @MargaretSiple-NOAA for steering this through.

**Comment by Selbosh** [comment]
## Reviewer Response


#### Final approval (post-review)

- [x] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**



Estimated hours spent reviewing: 6

**Comment by MargaretSiple-NOAA** [comment]
Thank you @Selbosh !

 Hi @agricolamz -- have you had enough time to review the changes that Mark made? I want to make sure you're happy with the current version and get things wrapped up within the next few weeks. Thank you.

**Comment by agricolamz** [comment]
Hello, all, I am really sorry for keeping all of you waiting...

The package now is much better and user friendlier! Thank you @mpadge for doing it! Now it is clearly reports what is it doing, where the data is downloaded etc. Unfortunately, I can't say that I am happy with the current version (see some comments below), but I don't want to be a stubborn donkey: if you all OK with the package, it can move to the next stage. Here are some comments on the following version of the package: 

```
packageVersion("pkgmatch")

[1] ‘0.5.2.117
```

## Major comments

- I recently found [RWsearch](https://cran.r-universe.dev/RWsearch) that promises some search through packages... I guess it should be mentioned somewhere in the documentation. I also think that it make sense to compare functionality and algorithm that they use.

> Search by keywords in R packages, task views, CRAN, the web and display the results in the console or in txt, html or pdf files. [source](https://cran.r-universe.dev/RWsearch/doc/manual.html#RWsearch-package)

- I think there should be a track for an archived packages. I run the following code:

```
library(pkgmatch)

"
Linguistic package for phonetic research.
" |> 
  pkgmatch_similar_pkgs(corpus = "ropensci")
```
and got

```
[1] "lingtypology"  "phonfieldwork" "pangoling" "mapmetadata"  "jstor"
```

Overall, I'm happy: I'm the author of `lingtypology` and `phonfieldwork` and they are indeed linguistic packages (but only the last one is for phonetic research). `pangoling` --- is a mixture of LLMs and linguistics. `mapmetadata` is not linguistic. But the `jstor` package (and this is why I'm writing it!) is archived at March. But why the function got it then?

- I tried to do a silly thing and try to input just `1`, and got an error that I kind of can't make any sense of.

```
pkgmatch_similar_pkgs(input = "1", corpus = "ropensci")

Error in get_pkg_text_local(pkg_name) : fs::dir_exists(path) is not TRUE
```

Funny that this can be fixed if you have `\n` in the `input`.

```
pkgmatch_similar_pkgs(input = "1\n", corpus = "ropensci")
[1] "BaseSet" "CRediTas" "CoordinateCleaner" "DataPackageR" "DataSpaceR"
```

It appears to be sensitive to spaces too:

```
pkgmatch_similar_pkgs(input =  "phonetic",
                      corpus = "ropensci")

Error in get_pkg_text_local(pkg_name) : fs::dir_exists(path) is not TRUE

pkgmatch_similar_pkgs(input =  "phonetic ", # Hi, I'm a space at the end!
                      corpus = "ropensci")

[1] "phonfieldwork" "BaseSet" "CRediTas" "CoordinateCleaner" "DataPackageR" 
```

I also want to mention that if you put rubbish into the `input` variable you will get those five packages we have seen before: `BaseSet`, `CRediTas`, `CoordinateCleaner`, `DataPackageR`, `DataSpaceR`. And this leads me to my next point.

- I think it make sense for both functions of the `pkgmatch` package to through something different, when the result package/function sorting is the same as default alphabetic order. This clearly indicates that package didn't make anything useful (and it is not the packages' fault).

- The argument `browse` return an error opening just one browser window/tab with the first package.
```
pkgmatch_similar_pkgs("1 ", corpus = "ropensci", browse = TRUE)

Error in vapply(urls, utils::browseURL, integer(1L)) : 
  values must be length 1,
 but FUN(X[[1]]) result is length 0
```

- *The shoemaker's children go barefoot*. I tried to find what is the similar packages for `pkgmatch` and `pkgmatch` is ... on the second place.

```
pkgmatch_similar_pkgs ("Search by keywords in R packages, rOpenSci, CRAN", corpus = "ropensci")
[1] "internetarchive" "pkgmatch" "ckanr" "rentrez" "codemetar"
```

I am really glad that when I asked the same question with CRAN packages I got:

```
pkgmatch_similar_pkgs ("Search by keywords in R packages, rOpenSci, CRAN", corpus = "cran")
[1] "pkgmatch" "RWsearch" "packagefinder" "serpstatr" "discoverableresearch"
```

So the package that I suggested before running the code makes sense! Your package tells you to look into it too! `packagefinder` looks promising (and was published in 2023, so I guess it should be mentioned in docs): the goal is the same, but some tools are different. I'd suggest to ask the author to review the `pkgmatch`, but it is probably too late. BTW, I think it is a good idea for rOpenSci: your package can easily suggest authors of matched packages as reviewers (after the manual check of relevance, of course).

- Since I found some heuristic for checking your package in the previous bullets I just applied them to the `pkgmatch_similar_fns()` function. The results are similar: if you put some rubbish, it will return the basic order. Unfortunately, if I put an example from `pkgmatch` documentation it doesn't return correct `pkgmatch` functions:

```
pkgmatch_similar_fns("Identify best match for an R function")
[1] "ritis::search_anymatch" "treebase::have_branchlength" "mapmetadata::user_categorisation_loop" "Athlytics::plot_pbs" "MODIStsp::get_yeardates"

pkgmatch_similar_fns("Find R packages matching an input")
[1] "coder::cols" "pkgmatch::pkgmatch_similar_fns" "PostcodesioR::postcode_autocomplete" "taxize::iucn_getname" "ckanr::package_search"
```

I wrote earlier that I am not happy with the current version of the package: the motivation for me writing it is the wiered behavior of the `pkgmatch_similar_fns()` function.

## Smaller issues

- Don't you want to allow users to use capital letters in the `corpus` argument? Right now when I type `pkgmatch_similar_pkgs("...", corpus = "CRAN")`, I get an error that I should type `cran` in lower case. An error message clearly suggest, what should I do, but I'd suggest to include some small fix here.
- 'A tool that uses language models' is the first statement in the package description, however I think that this statement is not true anymore.
- I have some doubts about code styling in the following chunk  `bm25 <- package <- NULL`. But may be it is metter of taste.
- You still have spaces between function names and the brackets, e.g. `foo ()` instead of `foo()`, even though we talked it through earlier.

**Comment by MargaretSiple-NOAA** [comment]
Thank you @agricolamz  for these productive additions; I agree with many of them. Hopefully a lot of these should be relatively easy:
- Remove archived packages from search results (or label them as archived)
- incorporate some early warnings/errors for inputs that won't work, e.g., numeric input for `pkgmatch_similar_pkgs()`
- incorporate some flexibility in character strings, as in his example with `pkgmatch_similar_pkgs(input =  "phonetic",...`. Users should get the same result a string whether or not it has leading or trailing spaces
- build in ux-friendly flexibility like using `cran` if a user enters `CRAN`. In general, I think the least case-dependent it is the better.

@agricolamz , let me know if I didn't capture any of those properly. The remaining three bullets of the 'Smaller issues' section I think are stylistic enough that if Mark feels strongly about them, they can stay in the package.

I remember a conversation we had at the beginning about the order of the search results and I have to say it has stuck with me; maybe because in the testing process we want to be able to recreate certain results reliably based on the inputs. I think this is a style thing, so I will not harp on it, but I do think it would be nice to have some prioritization in the results returned. If not that, then at least a clause within the function that will prevent `pkgmatch_similar_pkgs()` from returning the package itself if someone puts in a package name. 

I am at sea right now, so please forgive me if this response is a little loopier than my usual. And thank you again all for your contributions here.

**Comment by agricolamz** [comment]
Yeah, I agree, that most of my comments are easy to fix! 

Have a fruitful fieldwork!

**Comment by mpadge** [comment]
Thanks @mcsiple and @agricolamz - and no worries at all about being a "stubborn donkey" 😆 , all of you insights are only making the package better! I'll report back here as soon as I've implemented your suggestions.

**Comment by mpadge** [comment]
Thanks again @agricolamz for your always helpful insights, and to @mcsiple for the useful summary of actionable points. I've implemented the following updates in response.

>   Remove archived packages from search results (or label  them as archived)

Done. That was a complete oversight on my part, but now implemented [here](https://github.com/ropensci-review-tools/pkgmatch/commit/72b332226ffd664ffe972f64e49bccc9221a648a). Thanks for picking that up!

>  incorporate some early warnings/errors for inputs that won't work, e.g., numeric input for `pkgmatch_similar_pkgs()`

I'm similarly very grateful for you having picked this up. The main issue was that all inputs are passed through tokenizers, and so I've now implemented a catch for when no useful tokens are returned. You can see it in action in [this test](https://github.com/ropensci-review-tools/pkgmatch/blob/64c24ec8e580097351a84643b5515513122508be/tests/testthat/test-similar.R#L193-L215). Previous behaviour passed on through with no matchable tokens extracted from input, and so always returned generic and identically meaningless results. These updates should now fix that by erroring as soon as it's clear that the input has no useable tokens.

There's also a decision branch to detect whether inputs are supposed to be package names. That was also unreliable, which was why "phonetic" differed from "phonetic " - without space, it was detected as a package name, but with space as plain text. That's also fixed, so single words are only identified as package names if they match packages installed on local systems. (Because no matching can be performed on non-installed packages anyway.)

Note nevertheless that matching is still statistical, so even a phrase almost but not quite identical to, say, a package "_Description_" field might not necessarily return that package in first place. The primary purpose of all of this is to have a maximal probability of surfacing desired results within the top few. Your examples demonstrate this nicely:


``` r
pkgmatch_similar_fns("Find R packages matching an input")
#> [1] "coder::cols"                         "pkgmatch::pkgmatch_similar_fns"     
#> [3] "PostcodesioR::postcode_autocomplete" "taxize::iucn_getname"               
#> [5] "ckanr::package_search"
pkgmatch_similar_fns ("Identify best match for an R function")
#> [1] "ritis::search_anymatch"               
#> [2] "treebase::have_branchlength"          
#> [3] "mapmetadata::user_categorisation_loop"
#> [4] "Athlytics::plot_pbs"                  
#> [5] "MODIStsp::get_yeardates"
```

The first one works fine, even though `pkgmatch_similar_fns` is second and not first. The second one, however, doesn't even include the package, which is indeed unexpected and perhaps annoying. But if you tweak the phrasing just a little but, you see this:

``` r
pkgmatch_similar_fns ("Identify best matching R functions")
#> [1] "taxize::gna_verifier"           "pkgmatch::pkgmatch_similar_fns"
#> [3] "nlrx::util_get_os"              "nlrx::util_create_lhs"         
#> [5] "tabulapdf::extract_areas"
```

In this case, that happens because the token of "functions" remains as "functions", and "matching" remains as "matching". So the inputs get treated distinctly, with different results. That just one consequence of working with tokens, and for the purposes of this package, it is unavoidable. The only robust way to avoid it would be to use large language models to gauge embedding distances between tokens, and collapse any sets above some threshold similarity. But then we'd be back to the whole LLM approach which has already been shown via #254 to be less reliable than present approach.



> incorporate some flexibility in character strings, as in his example with `pkgmatch_similar_pkgs(input =  "phonetic",...`. Users should get the same result a string whether or not it has leading or trailing spaces

Fixed as explained in previous comment

> build in ux-friendly flexibility like using `cran` if a user enters `CRAN`. In general, I think the least case-dependent it is the better.

All done now with as much flexibility as possible. An initial letter will generally match, because "**r**opensci", "**c**ran", and "**b**ioc" are fortunately distinct. And all are now independent of case.

----

## Updates to comments from @mcsiple:
 
> I remember a conversation we had at the beginning about the order of the search results and I have to say it has stuck with me; maybe because in the testing process we want to be able to recreate certain results reliably based on the inputs. I think this is a style thing, so I will not harp on it, but I do think it would be nice to have some prioritization in the results returned. If not that, then at least a clause within the function that will prevent `pkgmatch_similar_pkgs()` from returning the package itself if someone puts in a package name.

Everything should now be entirely deterministically reproducible given the same input and the same (_not updated!_) local data for the corpora. Previous non-deterministic results were because of the LLM outputs, which have now been entirely removed,

And I understand the idea behind exclusion of the package itself, but I find myself using the function with `browse = TRUE` as an easy way to open the CRAN page of any package. More importantly, self-exclusion is tricker than it might superficially seem, and would be prone to inconsistent behaviours. There are a few factors at play there, but foremost is that matches are based on [tokenized representations](https://docs.ropensci.org/tokenizers) of all inputs, including package names. So self-exclusions immediately require a measure of how similar a token has to be to the actual name. And that is already subjective. That potential subjectivity then also cascades through a few further pathways such that it would become very error prone, and increase maintenance difficulty here.

So I hope it's alright with you, but for the main positive reason i wrote first, functions will always identify named packages. And yes, with the obvious caveat that `_similar_pkgs` is then, strictly speaking, a misnomer :laughing:

---

## Further updates to comments from @agricolamz 

Hopefully the initial section covers most of your concerns. A couple of additional points:

- The ~~_Prior Art_~~ _Similar Tools_ section has been [expanded to include `packagefinder`](https://github.com/ropensci-review-tools/pkgmatch/commit/5ce76e29a27277ad47f2b9023ddc52c488450d34), as you suggested, as well as a couple more tools.
- Remaining references to "a tool that uses language models" have been removed; thanks for picking those up!
- Your suggestion to use this tool to identify potential reviewers was a strong motivating case for our development of this package - great that you can also see application there!
- The code styling you have doubts about (`bm25 <- package <- NULL`) is preceded with the explanatory comment,
  ``` r
  # Suppress no visible binding notes:
  ```
  I strongly prefer that approach over defining global NULL variables, because global definitions quickly end up obsolete, and are impossible to accurately maintain. This keeps definitions as close as possible to where those notes are otherwise triggered.
- Finally, and with due apology, the spaces everywhere including function names are described in [CONTRIBUTING.md](https://github.com/ropensci-review-tools/pkgmatch/blob/main/CONTRIBUTING.md#code-style). In short: I maintain that,
  ``` r
  a <- this(is(a(function())))
  ```
  is simply more difficult for a human to visually parse than,
  ``` r
  a <- this (is (a (function ())))
  ```
  or,
  ``` r
  a <- this( is( a( function())))
  ```
  It doesn't matter where the spaces go, but having them somewhere is better than nowhere. In my opinion.

  ---

  Thanks both for another round of productive and helpful improvements. Hopefully nearly done now ...

**Comment by agricolamz** [comment]
@mpadge, thank you for your response! I'm happy with it and can go all official!

## Reviewer Response

#### Final approval (post-review)

- [x] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**

Estimated hours spent reviewing: 9

**Comment by mpadge** [comment]
Thanks so much @agricolamz, especially for bearing with me over such a long time frame. I'm happy for you that you're finally free of this, but have really appreciated all the excellent suggestions you've made. Your insights really have improved the package enormously! A final question: Would you agree to being listed as a reviewer in the package's DESCRIPTION file? A simple :+1: will suffice to affirm.

**Comment by agricolamz** [comment]
Thanks @mpadge, I am happy that I could provide a useful suggestions! Yes, sure, you can list me.

**Comment by MargaretSiple-NOAA** [comment]
Thanks so much @agricolamz -- I appreciate both of your inputs on this package. Going to move it to final approval.

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/671#issuecomment-4676164886 time 9

**Comment by MargaretSiple-NOAA** [comment]
@ropensci-review-bot approve pkgmatch

**Comment by mpadge** [comment]
Thanks so much @mcsiple for bearing with this extraordinarily long review process 🙏  And please note that the package will remain where it is, in github.com/ropensci-review-tools, rather than being transferred to `/ropensci/`. All tracking mechanisms for that are [already in place](https://github.com/ropensci/roregistry/blob/ffe02c98e71c6c8c4ba5a5f94efc2f659bdf8629/info/not_transferred.json#L56-L59). With that, I think we're finally done here 🎉 

And @Selbosh Are you also happy to be listed as an official reviewer in the package's `DESCRIPTION' file?
